demos/device_shadow_for_aws/shadow_demo_main.c

```diff
@@ -581,191 +595,214 @@ int RunDeviceShadowDemo( bool awsIotMqttMode,
+            if( xDemoStatus == pdPASS )
             {
-                LogError( ( "Failed to unsubscribe the topic %s",
-                            SHADOW_TOPIC_STRING_UPDATE_ACCEPTED( THING_NAME ) ) );
+                xDemoStatus = UnsubscribeFromTopic( &xMqttContext,
+                                                    SHADOW_TOPIC_STRING_UPDATE_REJECTED( THING_NAME ),
+                                                    SHADOW_TOPIC_LENGTH_UPDATE_REJECTED( THING_NAME_LENGTH ) );
+
+                if( xDemoStatus != pdPASS )
+                {
+                    LogError( ( "Failed to unsubscribe the topic %s",
+                                SHADOW_TOPIC_STRING_UPDATE_REJECTED( THING_NAME ) ) );
+                }
             }
-        }
 
-        if( xDemoStatus == pdPASS )
-        {
-            xDemoStatus = UnsubscribeFromTopic( &xMqttContext,
-                                                SHADOW_TOPIC_STRING_UPDATE_REJECTED( THING_NAME ),
-                                                SHADOW_TOPIC_LENGTH_UPDATE_REJECTED( THING_NAME_LENGTH ) );
+            /* The MQTT session is always disconnected, even there were prior failures. */
+            xDemoStatus = DisconnectMqttSession( &xMqttContext, &xNetworkContext );
 
-            if( xDemoStatus != pdPASS )
+            /* This demo performs only Device Shadow operations. If matching the Shadow
+             * MQTT topic fails or there are failure in parsing the received JSON document,
+             * then this demo was not successful. */
+            if( ( xUpdateAcceptedReturn != pdPASS ) || ( xUpdateDeltaReturn != pdPASS ) )
             {
-                LogError( ( "Failed to unsubscribe the topic %s",
-                            SHADOW_TOPIC_STRING_UPDATE_REJECTED( THING_NAME ) ) );
+                LogError( ( "Callback function failed." ) );
+                xDemoStatus = pdFAIL;
             }
         }
 
-        /* The MQTT session is always disconnected, even there were prior failures. */
-        xDemoStatus = DisconnectMqttSession( &xMqttContext, &xNetworkContext );
+        /* Increment the demo run count. */
+        ulDemoRunCount++;
 
-        /* This demo performs only Device Shadow operations. If matching the Shadow
-         * MQTT topic fails or there are failure in parsing the received JSON document,
-         * then this demo was not successful. */
-        if( ( xUpdateAcceptedReturn != pdPASS ) || ( xUpdateDeltaReturn != pdPASS ) )
+        if( xDemoStatus == pdPASS )
         {
-            LogError( ( "Callback function failed." ) );
-            xDemoStatus = pdFAIL;
+            LogInfo( ( "Demo iteration %lu is successful.", ulDemoRunCount ) );
         }
-    }
+        /* Attempt to retry a failed iteration of demo for up to #SHADOW_MAX_DEMO_COUNT times. */
+        else if( ulDemoRunCount < SHADOW_MAX_DEMO_COUNT )
+        {
+            LogWarn( ( "Demo iteration %lu failed. Retrying...", ulDemoRunCount ) );
```

Comment:
Note that the occurrences of this error message should be counted in the CI system.

Reply:
Yea, Thanks. That is a good suggestion.